text/0000-unsafe-extern-blocks.md

```diff
@@ -0,0 +1,120 @@
+* On editions < 2024, you *may* write `unsafe extern`, or you can write an `extern` block without the `unsafe` keyword. Writing an `extern` block without the `unsafe` keyword is provided for compatibility only, and will generate a warning.
+* `unsafe extern` interacts with the `unsafe_code` lint, and a `deny` or `forbid` with that lint will deny or forbid the unsafe external block.
+
+Within an `extern` block is zero or more declarations of external functions and/or external static values.
+An extern function is declared with a `;` instead of a function body (similar to a method of a trait).
+An extern static value is also declared with a `;` instead of an expression (similar to an associated const of a trait).
+In both cases, the actual function body or value is provided by whatever external source (which is probably not even written in Rust).
+
+When an `unsafe extern` block is used, all declarations within that `extern` block *must* have the `unsafe` or `safe` keywords as part of their signature.
+The `safe` keyword is a contextual keyword, it is currently only used within `extern` blocks.
+
+If an `extern` block is used in an older edition without the `unsafe` keyword, declarations *cannot* specify `safe` or `unsafe`.
+Code must update to `unsafe extern` style blocks if it wants to make `safe` declarations.
+
+```rust
+unsafe extern {
+    // sqrt (from libm) can be called with any `f64`
+    pub safe fn sqrt(x: f64) -> f64;
+    
+    // strlen (from libc) requires a valid pointer,
+    // so we mark it as being an unsafe fn
+    pub unsafe fn strlen(p: *const c_char) -> usize;
+
+    // this function doesn't say safe or unsafe, so it defaults to unsafe
+    pub fn free(p: *mut core::ffi::c_void);
+    
+    pub safe static IMPORTANT_BYTES: [u8; 256];
+    
+    pub safe static LINES: SyncUnsafeCell<i32>;
+}
+```
+
+`extern` blocks are `unsafe` because if the declaration doesn't match the actual external function, or the actual external data, then it causes compile time Undefined Behavior (UB).
+
+Once they are unsafely declared, a `safe` item can be used outside the `extern` block as if it were any other safe function or static value declared within rust.
+The unsafe obligation of ensuring that the correct items are being linked to is performed by the crate making the declaration, not the crate using that declaration.
+
+Items declared as `unsafe` *must* still have a correctly matching signature at compile time, but they *also* have some sort of additional obligation for correct usage at runtime.
+They can only be used within an `unsafe` block.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The grammar of the langauge is updated so that:
+
+* Editions >= 2024 *must* prefix all `extern` blocks with `unsafe`.
+* Editions < 2024 *should* prefix `extern` blocks with `unsafe`, this is a warn-by-default compatibility lint when `unsafe` is missing.
```

Comment:
the intention was to make function declarations consistent with function definitions ("unsafe" are explicitly annotated), and "eventually" all unannotated extern function declaration would be "safe", so the extern block itself has to be marked as "unsafe" to specify that safety has been checked

i'm not convinced this is a valid reason to force `unsafe` on *every* `extern` block. at the very least, if the block contains *only* `unsafe` items, there is no reason mark to `extern` block itself again, it is like `unsafe { unsafe { expr; } }` is an `unused_unsafe` warning.

i'm also not convinced that "eventually" killing the `safe` context keyword is a proper path to take, since an FFI fn declaration is a blackbox allowed to do any nasty thing, unlike a fn definition which `rustc` can inspect into the implementation to mechanically check its safety.

Reply:
We've added the word "eventually" to this section as a clarification about this warning.  The lang team is conscious of not creating too much noise from new warnings and will act carefully here during stabilization of any changes.

In terms of the rationale for these changes, see in particular [this comment](https://github.com/rust-lang/rfcs/pull/3484#issuecomment-1758275493) from **pnkfelix**.